### PR TITLE
perf: speed up search_session (columnar ingest, single index build, vectorized most-recent-belief selection)

### PR DESCRIPTION
**Pull request by Flix6x**
## What

Three independent speedups on the `search_session` hot path, one commit each, plus a committed benchmark script (`dev/bench_search_session.py`, 100k beliefs, 500 events x 100 horizons x 2 sources against the test database); a subset scenario returning 10% of the events is included for a more typical read size).

| scenario | before | after | change |
|---|---|---|---|
| plain search (100k rows returned) | 1491 ms | 1242 ms | **-17%** |
| plain search, 10% event window (10k rows returned) | 193 ms | 184 ms | **-5%** |
| `most_recent_beliefs_only` (SQL path) | 126 ms | 125 ms | - |
| most recent + `beliefs_before` (pandas fallback) | 19,866 ms | 565 ms | **-97%** |

Per-commit contribution (min of 3 reps, plain / fallback):

| commit | plain | fallback |
|---|---|---|
| baseline (main) | 1491 ms | 19,866 ms |
| build the index once, skip re-sorting on tz conversion | 1438 ms | 19,840 ms |
| skip needless ORM row processing | 1242 ms | 19,643 ms |
| vectorize `select_most_recent_belief` | 1265 ms | **565 ms** |

## How

1. **One index build, one sort**: `search_session` computes `belief_time` as a column (same formula as the `belief_times` property) before constructing the `BeliefsDataFrame`, so the final multi-index is built and sorted once instead of being rebuilt (deep copy + full re-sort) three times (horizon→time conversion + two timezone conversions). Timezone conversion of a tz-aware index level now converts the unique level values via `set_levels` — tz conversion preserves instants, so codes and sort order cannot change. `replace_multi_index_level` takes a shallow copy (it never mutates column data) and gains a `sort` parameter for order-preserving replacements.
2. **Core-connection fetch**: the beliefs query is executed on the Core connection (with an explicit flush to preserve autoflush semantics), skipping needless ORM row processing.
3. **Vectorized `select_most_recent_belief`**: the two groupby-apply passes (Python callable per belief group / per event-source group) are replaced with vectorized groupby transforms on the index level codes. Output now preserves input row order (call sites sort beforehand; the previous implementation returned groupby order).

Behavioral note: the timezone-conversion methods no longer implicitly sort a frame that was unsorted to begin with (no caller in this repo or FlexMeasures relies on that).

## Verification

- Full test suite passes (226 passed).
- FlexMeasures integration: ran FlexMeasures' search-path test modules (test_queries, test_time_series_services, test_sensor_data API tests, test_belief_charts — 89 tests) against this branch: all pass.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01AjAZFQmASY65gPmVtQAahG

### CODE REVIEW

**Comment by nhoening** [comment]
- What is the use case for the one impactful scenario ("most recent + beliefs_before (pandas fallback)")? I am not sure from the name - do we encounter this often in practice?
- Can we also test plain search with only a subset of the test data being returned (I believe now it returns all 100k lines, which is more of an edge case)?

**Comment by Flix6x** [comment]
Good questions!

**1. When does the fallback scenario occur in practice?**

The pandas fallback triggers when all three of these hold (see `most_recent_beliefs_only_incompatible_criteria` in `search_session`):

1. `most_recent_beliefs_only=True` (FlexMeasures' default for reads), and
2. a belief-time cutoff is set (`beliefs_before`/`beliefs_after`), and
3. the sensor's knowledge horizon function is *not* plain `ex_ante`/`ex_post` (because then the SQL can't compute belief times from horizons, so the most-recent subquery is skipped and everything is filtered in pandas).

In FlexMeasures terms, condition 3 means day-ahead market price sensors (`x_days_ago_at_y_oclock` — e.g. the toy-account day-ahead sensor and epex_da), and condition 2 is the `prior` parameter: `GET /sensors/data` with `prior`, and schedulers/reporters passing `beliefs_before` to reconstruct the state of knowledge at decision time. So the typical hit is **backtesting/simulating scheduling against day-ahead prices** — scheduling with an explicit `prior` fetches prices through exactly this path, once per scheduling job. That's why I think the 20s → 0.6s win is worth having: it's not the everyday live-scheduling path (no `prior` → SQL path), but it's the standard simulation setup, where it's also called many times in a row.

**2. Benchmarking a subset read**

Good point — added a `plain_subset` scenario (event window covering 10% of the events, so 10k of 100k rows are returned) in 8763a5c. Results (min of 3 reps):

| scenario | before (main) | after | change |
|---|---|---|---|
| plain, all 100k rows | 1,459 ms | 1,258 ms | -14% |
| plain, 10% window (10k rows) | 193 ms | 184 ms | -5% |

The gains on this branch scale with the number of rows *returned*, since they remove per-row ingest and whole-frame post-processing costs; at 10k rows those overheads are small in absolute terms and query+fetch dominates, so the subset read only improves modestly. The big wins here are large reads and the fallback path above. (Making small window reads faster is more about the SQL side — indexes and your materialized view work.)

I also updated the table in the PR description with the subset row.

**Comment by Flix6x** [comment]
Follow-up: #240 (stacked on this PR) fetches the timing columns as epoch microseconds, which roughly halves the remaining search time again — notably also for the 10%-window subset read discussed above (193 ms on main → 106 ms with both PRs), since per-row object materialization was exactly the cost that scaled with rows returned.

File: timely_beliefs/utils.py [hunk 1]
```diff
@@ -174,15 +177,17 @@ def replace_multi_index_level(
     # Construct new MultiIndex
     mux = pd.MultiIndex.from_arrays(new_index_values, names=new_index_names)
 
-    df = df.copy(deep=True)
+    # A shallow copy suffices: column data is never mutated here, and both
+    # sort_index() and reindex() return frames with their own (non-aliased) data
+    df = df.copy(deep=False)
```
[the remark this answers is not in the document]
Comment (Flix6x) [reply timely_beliefs/utils.py]: Good catch — fixed in 7d7ecc9 by restoring the deep copy (and doing the same in the new timezone conversion fast path, which had the same aliasing).

Rerunning `dev/bench_search_session.py` shows no measurable difference in any of the four scenarios, so the shallow copy was not where the speedup came from: that was building and sorting the index once, rather than skipping the data copy.
